vm/src/stdlib/ctypes/function.rs

```diff
@@ -1,24 +1,182 @@
+                } else {
+                    Err(vm.new_type_error("Expected a ctypes simple type".to_string()))
+                }
+            })
+            .collect::<PyResult<Vec<Type>>>()?;
+        let terminated = format!("{}\0", function);
+        let pointer: Symbol<'_, FP> = unsafe {
+            library
+                .get(terminated.as_bytes())
+                .map_err(|err| err.to_string())
+                .map_err(|err| vm.new_value_error(err))?
+        };
+        let code_ptr = CodePtr(*pointer as *mut _);
+        let return_type = match ret_type {
+            // TODO: Fix this
+            Some(_t) => {
+                return Err(vm.new_not_implemented_error("Return type not implemented".to_string()));
+            }
+            None => Type::c_int(),
+        };
+        let cif = Cif::new(args, return_type);
+        Ok(Function {
+            cif,
+            pointer: code_ptr,
+        })
+    }
+
+    pub unsafe fn call(
+        &self,
+        args: Vec<PyObjectRef>,
+        vm: &VirtualMachine,
+    ) -> PyResult<PyObjectRef> {
+        let args = args
+            .into_iter()
+            .map(|arg| {
+                if let Some(data) = arg.downcast_ref::<PyCSimple>() {
+                    dbg!(&data);
+                    todo!("HANDLE ARGUMENTS")
+                } else {
+                    Err(vm.new_type_error("Expected a ctypes simple type".to_string()))
+                }
+            })
+            .collect::<PyResult<Vec<Arg>>>()?;
+        // TODO: FIX return type
+        let result: i32 = unsafe { self.cif.call(self.pointer, &args) };
+        Ok(vm.ctx.new_int(result).into())
+    }
 }
 
 #[pyclass(module = "_ctypes", name = "CFuncPtr", base = "PyCData")]
+#[derive(PyPayload)]
 pub struct PyCFuncPtr {
-    pub _name_: String,
-    pub _argtypes_: AtomicCell<Vec<PyObjectRef>>,
-    pub _restype_: AtomicCell<PyObjectRef>,
-    _handle: PyObjectRef,
-    _f: PyRwLock<Function>,
+    pub name: PyRwLock<String>,
+    pub _flags_: AtomicCell<u32>,
+    // FIXME(arihant2math): This shouldn't be an option, setting the default as the none type should work
+    //  This is a workaround for now and I'll fix it later
+    pub _restype_: PyRwLock<Option<PyTypeRef>>,
+    pub handler: PyObjectRef,
+}
+
+impl Debug for PyCFuncPtr {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        f.debug_struct("PyCFuncPtr")
+            .field("name", &self.name)
+            .finish()
+    }
 }
 
-#[pyclass]
-impl PyCFuncPtr {}
+impl Constructor for PyCFuncPtr {
+    type Args = FuncArgs;
+
+    fn py_new(_cls: PyTypeRef, args: Self::Args, vm: &VirtualMachine) -> PyResult {
+        let tuple = args.args.first().ok_or_else(|| {
+            vm.new_type_error("CFuncPtr() takes exactly 1 argument (0 given)".to_string())
+        })?;
+        let tuple: &Py<PyTuple> = tuple
+            .downcast_ref()
+            .ok_or(vm.new_type_error("Expected a tuple".to_string()))?;
```

Comment:
Oh, just one more thing. Let's do this here. We can more freely split FuncArgs

```suggestion
    type Args = (PyTupleRef, FuncArgs);

    fn py_new(_cls: PyTypeRef, (tuple, _args): Self::Args, vm: &VirtualMachine) -> PyResult {
```